core/drivers/lpc_uart.c

```diff
@@ -0,0 +1,125 @@
+// SPDX-License-Identifier: BSD-2-Clause
+/*
+ * Copyright (c) 2022, HiSilicon Limited
+ */
+
+#include <assert.h>
+#include <drivers/lpc_uart.h>
+#include <io.h>
+#include <keep.h>
+#include <mm/core_mmu.h>
+#include <util.h>
+
+static paddr_t chip_to_base(struct serial_chip *chip)
+{
+	struct lpc_uart_data *pd =
+		container_of(chip, struct lpc_uart_data, chip);
+
+	return io_pa_or_va(&pd->base, LPC_SIZE);
+}
+
+static void lpc_byte_read(paddr_t addr, uint8_t *data)
+{
+	uint32_t status = 0;
+	uint32_t cnt = 0;
+
+	io_write32(LPC_CMD_REG_OFFSET + addr, LPC_SINGLE_READ);
+
+	io_write32(LPC_OP_LEN_REG_OFFSET + addr, 1);
+	io_write32(LPC_ADDR_REG_OFFSET + addr, UART_BASE + UART_LSR);
+
+	io_write32(LPC_START_REG_OFFSET + addr, 1);
+
+	status = io_read32(LPC_IRQ_ST_REG_OFFSET + addr);
+	while (!(status & LPC_IRQ_ST_ON)) {
+		if (cnt > UART_SEND_LOOP_MAX)
+			return;
+		cnt++;
+		status = io_read32(LPC_IRQ_ST_REG_OFFSET + addr);
+	}
+
+	io_write32(LPC_IRQ_ST_REG_OFFSET + addr, LPC_IRQ_ST_ON);
+
+	if (io_read32(LPC_OP_STATUS_REG_OFFSET + addr) & LPC_IRQ_ST_ON)
+		*data = io_read32(LPC_RDATA_REG_OFFSET + addr);
+}
+
+static void lpc_byte_write(paddr_t addr, uint8_t data)
+{
+	uint32_t status = 0;
+	uint32_t cnt = 0;
+
+	io_write32(LPC_CMD_REG_OFFSET + addr, LPC_SINGLE_WRITE);
+	io_write32(LPC_OP_LEN_REG_OFFSET + addr, 1);
+	io_write32(LPC_WDATA_REG_OFFSET + addr, data);
+
+	io_write32(LPC_ADDR_REG_OFFSET + addr, UART_BASE + UART_THR);
+	io_write32(LPC_START_REG_OFFSET + addr, 1);
+
+	status = io_read32(LPC_IRQ_ST_REG_OFFSET + addr);
+	while (!(status & LPC_IRQ_ST_ON)) {
+		if (cnt > UART_SEND_LOOP_MAX)
+			return;
+		cnt++;
+		status = io_read32(LPC_IRQ_ST_REG_OFFSET + addr);
+	}
+
+	io_write32(LPC_IRQ_ST_REG_OFFSET + addr, LPC_IRQ_ST_ON);
+}
+
+static void lpc_uart_core_putc(paddr_t base, int ch)
+{
+	uint8_t var = '\0';
+	uint32_t i = 0;
+
+	for (i = 0; i < UART_SEND_LOOP_MAX; i++) {
+		lpc_byte_read(base, &var);
+		if ((var & LPC_RADTA_LEN) == LPC_RADTA_LEN)
+			break;
+	}
+
+	lpc_byte_write(base, ch);
+
+	for (i = 0; i < UART_SEND_LOOP_MAX; i++) {
+		lpc_byte_read(base, &var);
+		if ((var & LPC_RADTA_LEN) == LPC_RADTA_LEN)
+			break;
+	}
+}
+
+static void lpc_uart_putc(struct serial_chip *chip, int ch)
+{
+	paddr_t base = chip_to_base(chip);
+
+	lpc_uart_core_putc(base, ch);
+}
+
+static void lpc_uart_flush(struct serial_chip *chip __unused)
+{
+}
+
+static bool lpc_uart_have_rx_data(struct serial_chip *chip __unused)
+{
+	return true;
+}
+
+static int lpc_uart_getchar(struct serial_chip *chip __unused)
+{
+	return 0;
+}
+
+static const struct serial_ops lpc_uart_ops = {
+	.flush = lpc_uart_flush,
+	.getchar = lpc_uart_getchar,
+	.have_rx_data = lpc_uart_have_rx_data,
+	.putc = lpc_uart_putc,
```

Comment:
Sorry I forgot one thing. There is no need to provide stubs for `have_rx_data()` and `getchar()`, and with https://github.com/OP-TEE/optee_os/pull/5690 `flush()` should not be required either. Can you please remove them?

Reply:
I will check it now. thx

Reply:
Please remove the `= NULL` assignments. The C standard guarantees that the missing fields are initialized to zero.
And also remove `.flush` -- I will make sure this gets merged after #5690.
```C
static const struct serial_ops lpc_uart_ops = {
	.putc = lpc_uart_putc,
};
```

Reply:
I've removed this unused lines.